modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -4629,3 +4636,179 @@ def from_dataframe(cls, df: "ProtocolDataframe") -> "PandasDataframe":
         ErrorMessage.default_to_pandas(message="`from_dataframe`")
         pandas_df = from_dataframe_to_pandas(df)
         return cls.from_pandas(pandas_df)
+
+    def case_when(self, caselist):
+        """
+        Replace values where the conditions are True.
+
+        Parameters
+        ----------
+        caselist : list of tuples
+
+        Returns
+        -------
+        PandasDataframe
+        """
+        # For Dask the callables must be wrapped for each partition, otherwise
+        # the execution could fail with CancelledError.
+        single_wrap = Engine.get() != "Dask"
+        cls = type(self)
+        wrapper_put = self._partition_mgr_cls._execution_wrapper.put
+        if (
+            not single_wrap
+            or (remote_fn := getattr(cls, "_CASE_WHEN_FN", None)) is None
+        ):
+
+            def case_when(df, name, caselist):  # pragma: no cover
+                caselist = [
+                    tuple(
+                        (
+                            data.squeeze(axis=1)
+                            if isinstance(data, pandas.DataFrame)
+                            else data
+                        )
+                        for data in case_tuple
+                    )
+                    for case_tuple in caselist
+                ]
+                return pandas.DataFrame({name: df.squeeze(axis=1).case_when(caselist)})
+
+            if single_wrap:
+                cls._CASE_WHEN_FN = remote_fn = wrapper_put(case_when)
+            else:
+                remote_fn = case_when
+
+        name = self.columns[0]
+        use_map = single_wrap
+        is_trivial_idx = None
+        # Lists of modin frames: first for conditions, second for replacements
+        modin_lists = [[], []]
+        # Fill values for conditions and replacements respectively
+        fill_values = [True, None]
+        new_caselist = []
+        for case_tuple in caselist:
+            new_case = []
+            for data, modin_list, fill_value in zip(
+                case_tuple, modin_lists, fill_values
+            ):
+                if isinstance(data, cls):
+                    modin_list.append(data)
+                elif callable(data):
+                    if single_wrap:
+                        data = wrapper_put(data)
+                elif isinstance(data, pandas.Series):
```

Comment:
I'm not sure if it's efficient. In that case we will need to put all the objects into the remote storage, replace them with references and then pass the references to the remote functions. It may result in additional serialization, mere over, these objects are required for a single call only and will not be used anymore.

Reply:
> Wouldn't it be easier to convert to Modin objects at the top level?

I think Anatoly meant, wouldn't it be easier for us? :)

I don't think that we support non-distributed pandas object in any other modin's method. Most of the time we have an explicit check at the high level [verifying that we the function received a <i>modin</i> dataframe](https://github.com/modin-project/modin/blob/7c835a2761ede41d402c1febd29826c1d0b9512f/modin/pandas/general.py#L236-L239).

It's okay for this method to support pandas objects (since the time was already spent on it), but in the future, we shouldn't spend time on supporting such cases.